proposals/0047-rerouting.md

```diff
@@ -0,0 +1,186 @@
+<!--
+  Note: You are probably looking for `stage-1--discussion-template.md`!
+  This template is reserved for anyone championing an already-approved proposal.
+
+  Community members who would like to propose an idea or feature should begin
+  by creating a GitHub Discussion. See the repo README.md for more info.
+
+  To use this template: create a new, empty file in the repo under `proposals/${ID}.md`.
+  Replace `${ID}` with the official accepted proposal ID, found in the GitHub Issue
+  of the accepted proposal.
+-->
+
+- Start Date: 2024-04-17
+- Reference Issues: https://github.com/withastro/roadmap/issues/896
+- Implementation PR: 
+
+# Summary
+
+Programmatic control over routing.
+
+# Background & Motivation
+
+
+As astro grows in popularity, it is being used for more and more use cases. One of these is multitenant websites where a tenant may be given a subdomain. This proposal introduces a powerful primitive that allows astro users to build websites with complex routing requirements. Rather than a limitation, file-based routing is now a sensible default!
+
+# Goals
+
+- Defuse.
+- Make it unnecessary to rely on implicit routing rules
+- Introduce framework-designed 404 and 500 pages
+- Make middleware run for requests that don't match a route (with framework-designed error pages)
+
+# Non-Goals
+
+- Reroute to an external service.
+- Support for `functionPerRoute`
+- fallback rerouting
+
+# Detailed Design
+
+## Proposed APIs
+
+The rerouting will be exposed to Astro pages, Astro endpoints and middleware. This API is a function called `reroute` that will contain the following signature:
+
+```ts
+type Reroute = string | URL | Request
+const reroute: (payload: Reroute) => Promise<Response>
+
+reroute("/fr/hello");
+reroute(new URL("https://example.com"));
+reroute(new Request("https://localhost:8080", options));
+```
+
+Internally, the `reroute` **must** create a new `Request` when attempting to render the rerouted route. The creation of this new `Request` will vary based on the signature used.
+
+- Accepting a `string` allows to quickly reroute to a URL without too much hassle. When using a `string`, Astro will create a new `Request` with the new URL, and it will inherit all the data from the previous request.
+  > Astro won't do any particular check on the string. For example, it won't check for trailing slashes. 
+- Accepting a `URL` allows to create a more stable URL using the standard `URL` object. When using a `URL`, Astro will create a new `Request` with the new URL, and it will inherit all the data from the previous request.
```

Comment:
If you're looking for a use-case, providing a URL is nice when you are rerouting from a nested path because you can use a relative like, for ex. if you are at `/foo/bar/baz` and want to go to `/foo/bar/qux` you can do `reroute(new URL('./qux', Astro.url))`